### PR TITLE
#2641

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#2641 Indentation messed up for incomplete expressions 

Parse expression line we do in REPL in order to determine if it is complete. Based on the completeness drive smart indent and user indentation preservation.

### CODE REVIEW

File: src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs [hunk 1]
```diff
@@ -38,5 +39,18 @@ public static class AstRootExtensions {
             }
             return null;
         }
+
+        public static bool IsCompleteExpression(this AstRoot expressionAst) {
+            foreach (var error in expressionAst.Errors) {
+                if (error.ErrorType == ParseErrorType.CloseCurlyBraceExpected ||
+                    error.ErrorType == ParseErrorType.CloseBraceExpected ||
+                    error.ErrorType == ParseErrorType.CloseSquareBracketExpected ||
+                    error.ErrorType == ParseErrorType.FunctionBodyExpected ||
```
Comment (int19h) [inline src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs]: It seems that this should apply to all composite statements that have a mandatory body - e.g. `if`, `while` etc. Do we do that already? If so, this enum member should probably be renamed to `BodyExpected`.

Comment (MikhailArkhipov) [inline src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs]: This is code we used to use in REPL to show continuation. But no, it does not currently detect missing body on scope-less 'if' or similar.

Comment (MikhailArkhipov) [inline src/R/Core/Impl/AST/Extensions/AstRootExtensions.cs]: #2655

File: src/R/Core/Impl/Formatting/RFormatter.cs [hunk 2]
```diff
@@ -651,7 +653,9 @@ public class RFormatter {
             int end = _tokens.IsEndOfStream() ? _textProvider.Length : _tokens.CurrentToken.Start;
 
             string text = _textProvider.GetText(TextRange.FromBounds(start, end));
-            if (!preserveUserIndent && string.IsNullOrWhiteSpace(text)) {
+            bool lineBreakBeforeToken = _tokens.IsLineBreakAfter(_textProvider, _tokens.Position - 1);
+
+            if (lineBreakBeforeToken && !preserveUserIndent && string.IsNullOrWhiteSpace(text)) {
```
Comment (MikhailArkhipov) [inline src/R/Core/Impl/Formatting/RFormatter.cs]: Yep

File: src/R/Core/Impl/Formatting/FormattingScope.cs [hunk 3]
```diff
@@ -17,7 +17,8 @@ internal sealed class FormattingScope : IDisposable {
         private readonly TextBuilder _tb;
         private readonly int _previousIndentLevel;
 
-        public int CloseBracePosition { get; private set; } = -1;
+        public RToken OpenBraceToken { get; set; }
+        public int CloseBracePosition { get; } = -1;
```
Comment (MikhailArkhipov) [inline src/R/Core/Impl/Formatting/FormattingScope.cs]: It is set in ctor so I guess initial value is not needed

**Comment by AlexanderSher** [comment]
👍

**Pull request merged**